File: src/main/java/dev/reeve/mazelib/MazePosition.java
package dev.reeve.mazelib;

import lombok.Getter;
import lombok.NonNull;

import java.util.Objects;

@Getter
public class MazePosition implements Cloneable {
	private final int x;
	private final int y;
	
	public MazePosition(int x, int y) {
		this.x = x;
		this.y = y;
	}
	
	public MazePosition inDirection(@NonNull MazeDirection direction) {
		return new MazePosition(x + direction.getX(), y + direction.getY());
	}
	
	@Override
	public MazePosition clone() {
		return new MazePosition(x, y);
	}
	
	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof MazePosition)) return false;
		MazePosition that = (MazePosition) o;
		return x == that.x && y == that.y;
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(x, y);
	}
	
	@Override
	public String toString() {
		return "MazePosition{" +
				"x=" + x +
				", y=" + y +
				'}';
	}
}

File: src/main/kotlin/dev/reeve/mazelib/MazePosition.kt
package dev.reeve.mazelib

import java.util.Objects

class MazePosition(val x: Int, val y: Int) : Cloneable {
	fun inDirection(direction: MazeDirection): MazePosition {
		return MazePosition(x + direction.x, y + direction.y)
	}
	
	public override fun clone(): MazePosition {
		return MazePosition(x, y)
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is MazePosition) return false
		return x == other.x && y == other.y
	}
	
	override fun hashCode(): Int {
		return Objects.hash(x, y)
	}
	
	override fun toString(): String {
		return "MazePosition{" +
				"x=" + x +
				", y=" + y +
				'}'
	}
}